### PR TITLE
[23.05] Backport CONFIG_WERROR fixes

### PR DESCRIPTION
**Pull request by ynezz**
In commit b2d1eb717b6 ("generic: 5.15: enable Werror by default for kernel compile") `CONFIG_WERROR=y` was enabled and all warnings/errors reported with GCC 12 were fixed.

Keeping this in sync with past/future GCC/kernel versions and config option combinations is going to be uphill battle, so lets backport the `KERNEL_WERROR` config option, which keeps it enabled by default for tested/known working combinations and on buildbots. On the CI its [enabled explicitly](https://github.com/openwrt/actions-shared-workflows/blob/main/.github/workflows/reusable_build.yml#L446-L450).

In other words, it should keep the current behavior, but makes it configurable. 

References: #12687 #12622 #12734 #12713 #15064 

Cc: @robimarko @dangowrt @Ansuel

### CODE REVIEW

**Comment by ynezz** [comment]
@Ansuel thanks, that was fast :)

BTW I've enabled [vigilant mode](https://docs.github.com/en/authentication/managing-commit-signature-verification/displaying-verification-statuses-for-all-of-your-commits#about-vigilant-mode) so it now shows 33612fdb92 as `Partially verified` because I'm the author, but there is no signature. So perhaps GitHub issue, in such cases, they should probably not allow `Allow edits and access to secrets by maintainers`  and contributor would need to rebase and sign again? Just something new to be aware of.
